### PR TITLE
Deprecated imports in docstring examples in `io.harwell_boeing` and `optimize.nonlin`

### PR DESCRIPTION
**Pull request by ev-br**
Just a start on gh-16531

### CODE REVIEW

File: scipy/optimize/_nonlin.py [hunk 1]
```diff
@@ -1338,8 +1338,8 @@ class KrylovJacobian(Jacobian):
         Note that you can use also inverse Jacobians as (adaptive)
         preconditioners. For example,
 
-        >>> from scipy.optimize.nonlin import BroydenFirst, KrylovJacobian
-        >>> from scipy.optimize.nonlin import InverseJacobian
+        >>> from scipy.optimize import BroydenFirst, KrylovJacobian
+        >>> from scipy.optimize import InverseJacobian
```
Comment (WarrenWeckesser) [inline scipy/optimize/_nonlin.py:1342]: Do these imports work for you?  When I try this (with SciPy version'1.10.0.dev0+1373.318d8c6'), none of these names can be imported from `scipy.optimize`.

**Comment by ev-br** [comment]
Hmm, not sure what's going on with the InverseJacobian, BroydenFirst, and KrylovJacobian : importing them from `scipy.optimize.nonlin` triggers a deprecation warning and a recommendation to use a main `scipy.optimize` namespace. However, none of these can be imported either:

```
In [1]: from scipy.optimize.nonlin import InverseJacobian
<ipython-input-1-0e387586be55>:1: DeprecationWarning: Please use `InverseJacobian` from the `scipy.optimize` namespace, the `scipy.optimize.nonlin` namespace is deprecated.
  from scipy.optimize.nonlin import InverseJacobian

In [2]: from scipy.optimize import InverseJacobian
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Input In [2], in <cell line: 1>()
----> 1 from scipy.optimize import InverseJacobian

ImportError: cannot import name 'InverseJacobian' from 'scipy.optimize' (/home/br/repos/scipy/scipy/build-install/lib/python3.8/site-packages/scipy/optimize/__init__.py)
```

Trying in ipython:

```
In [1]: import scipy.optimize as so

In [2]: any(['Krylov' in _ for _ in dir(so)])
Out[2]: False

In [3]: import scipy.optimize._nonlin as sno

In [4]: any(['Krylov' in _ for _ in dir(sno)])
Out[4]: True
```

So I'm not sure what is the intent: keep them in `optimize.nonlin` or in the top-level `optimize`. Or are they supposed to really be private (a deprecation period, maybe?)
And I'm surprised these things being not importable is not caught by real tests, are these completely untested?

Not sure who's the main optimize resident these days. Mind if I ask you @andyfaff ?

File: scipy/optimize/_nonlin.py (hunk 1 above)
Comment (ev-br) [reply scipy/optimize/_nonlin.py:1342]: https://github.com/scipy/scipy/pull/16532#issuecomment-1173112142 comments crossed

**Comment by WarrenWeckesser** [comment]
It looks like this goes back to https://github.com/scipy/scipy/pull/14966.  Ping @czgdp1807 and @rgommers.  I suspect the question will be if those names should also be added to the `scipy.optimize` namespace.

**Comment by rgommers** [comment]
This goes back to commit 0c9c174c3 from 2010, which added `InverseJacobian` to `nonlin.py` but didn't expose it in `optimize`. Since that commit also set `.. currentmodule:: scipy.optimize` in the `optimize.nonlin` docstring, it is pretty clear that `optimize.nonlin` was never really public.

So now the question is whether `InverseJacobian` should be public yes or no. I'm not sure to be honest. WDYT?

**Comment by ev-br** [comment]
While I don't have a horse in this race, I'd say let's add them to the main optimize namespace and be done with them. As a guess they are not very widely used, but in those rare cases when they are needed, having them available is really helpful (_no trouble unless there is trouble, and then maybe a whole lot of trouble_) .

**Comment by rgommers** [comment]
> While I don't have a horse in this race, I'd say let's add them to the main optimize namespace and be done with them.

+1 from me. Do you prefer to add that to this PR, or do we need a separate PR?

**Comment by ev-br** [comment]
> +1 from me. Do you prefer to add that to this PR, or do we need a separate PR?

No need for a separate PR I'd say; have updated this PR with the three names to the scipy.optimize namespace.

**Review by rgommers — APPROVED** [review]
LGTM now, and CI failures are unrelated. In it goes. Thanks @ev-br

**Pull request merged**